### PR TITLE
Fix C# connection to correctly mark message as sent

### LINKED ISSUE #2722: csharp/Ice/retry failure (ubuntu)
from https://github.com/zeroc-ice/ice/actions/runs/10709606416/job/29694630089?pr=2720

```
*** [51/308] Running csharp/Ice/retry tests ***
[ running client/server test - 09/04/24 21:38:15 ]
- Config: ssl
(dotnet /home/runner/work/ice/ice/csharp/test/Ice/retry/msbuild/server/net8.0/server.dll --Ice.Default.Host=127.0.0.1 --Test.BasePort=14000 --Ice.Warn.Connections=1 --Ice.Default.Protocol=ssl --Ice.IPv6=0 --IceSSL.*** --IceSSL.DefaultDir=/home/runner/work/ice/ice/certs --IceSSL.CAs=cacert.pem --IceSSL.VerifyPeer=2 --IceSSL.CertFile=server.p12 --Ice.ThreadPool.Server.Size=1 --Ice.ThreadPool.Server.SizeMax=3 --Ice.ThreadPool.Server.SizeWarn=0 --Ice.PrintAdapterReady=1 --Ice.Trace.Network=2 --Ice.Trace.Retry=1 --Ice.Trace.Protocol=1 --Ice.LogFile=/home/runner/work/ice/ice/csharp/test/Ice/retry/server-090424-2138.log)
(dotnet /home/runner/work/ice/ice/csharp/test/Ice/retry/msbuild/client/net8.0/client.dll --Ice.Default.Host=127.0.0.1 --Test.BasePort=14000 --Ice.Warn.Connections=1 --Ice.Default.Protocol=ssl --Ice.IPv6=0 --IceSSL.*** --IceSSL.DefaultDir=/home/runner/work/ice/ice/certs --IceSSL.CAs=cacert.pem --IceSSL.VerifyPeer=2 --IceSSL.CertFile=client.p12 --Ice.Trace.Network=2 --Ice.Trace.Retry=1 --Ice.Trace.Protocol=1 --Ice.LogFile=/home/runner/work/ice/ice/csharp/test/Ice/retry/client-090424-2138.log)
testing stringToProxy... ok
testing checked cast... ok
calling regular operation with first proxy... ok
calling operation to kill connection with second proxy... ok
calling regular operation with first proxy again... ok
calling regular AMI operation with first proxy... ok
calling AMI operation to kill connection with second proxy... value = 1, expected = 0
failed:
   at Test.TestHelper.fail(String message, String detailMessage) in /home/runner/work/ice/ice/csharp/test/TestCommon/TestHelper.cs:line 183
   at Test.TestHelper.TestTraceListener.Fail(String message, String detailMessage) in /home/runner/work/ice/ice/csharp/test/TestCommon/TestHelper.cs:line 35
   at System.Diagnostics.TraceInternal.Fail(String message, String detailMessage)
   at System.Diagnostics.Debug.Fail(String message, String detailMessage)
   at Ice.retry.Instrumentation.testEqual(Int32& value, Int32 expected) in /home/runner/work/ice/ice/csharp/test/Ice/retry/Instrumentation.cs:line 147
   at Ice.retry.Instrumentation.testRetryCount(Int32 expected) in /home/runner/work/ice/ice/csharp/test/Ice/retry/Instrumentation.cs:line 156
   at Ice.retry.AllTests.allTests(TestHelper helper, Communicator communicator, Communicator communicator2, String rf) in /home/runner/work/ice/ice/csharp/test/Ice/retry/AllTests.cs:line 91
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.ExecutionContextCallback(Object s)
   at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Runtime.CompilerServices.AsyncTaskMethodBuilder`1.AsyncStateMachineBox`1.MoveNext(Thread threadPoolThread)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
saved /home/runner/work/ice/ice/csharp/test/Ice/retry/client-090424-2138.log
saved /home/runner/work/ice/ice/csharp/test/Ice/retry/server-090424-2138.log

```

### PR DESCRIPTION
**Pull request by pepone**
This PR fixes C# Connection to correctly mark messages as sent, once that startWrite handles the complete buffer.

This bug was introduced in  #2270, probably and oversight because of the subtle difference between C# and C++ startWrite calls.

Here to warrantee at-most once semantics we have to mark the request as sent, once that startWrite is has consumed the complete buffer.

For C++ this is the return value of `startWrite`, for C# the return value is whether or not the write call completed synchronously and there is a separate out parameter that indicates whether or not the call consumed the complete buffer.

This must be back ported to 3.7, but the issue doesn't affect 3.7.10.

Fixes #2722

### CODE REVIEW

File: csharp/src/Ice/ConnectionI.cs [hunk 1]
```diff
@@ -607,15 +607,23 @@ public override bool startAsync(int operation, Ice.Internal.AsyncCallback comple
...
 
-                        bool completed;
-                        if (_transceiver.startWrite(_writeStream.getBuffer(), completedCallback, this, out completed))
+                        bool messageWritten = false;
+                        bool completedSynchronously =
+                            _transceiver.startWrite(
+                                _writeStream.getBuffer(),
+                                completedCallback,
+                                this,
+                                out messageWritten);
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:616]: It's not clear to me why the C# startWrite has such a complicated signature compared to the C++ startWrite. In C++, we don't have this "completedSynchronously" return value. 

In any case, assuming this complicated C# API is warranted, we should:
 - rename the last parameter of Transceiver.startWrite. Having return value = completedSynchronously and out parameter = completed is too confusing.
 - consider returning a tuple (does this work with the old C# version we use on 3.7?)
 - name the "messageWritten" parameter => messageFullyWritten
 - add a real doc-comment on Transceiver.startWrite

File: csharp/src/Ice/ConnectionI.cs [hunk 2]
```diff
@@ -607,15 +607,23 @@ public override bool startAsync(int operation, Ice.Internal.AsyncCallback comple
...
+                        // If the startWrite call consumed the complete buffer, we assume the message is sent now for
+                        // at-most-once semantics.
+                        if (messageWritten && _sendStreams.Count > 0)
```
Comment (bernardnormier) [inline csharp/src/Ice/ConnectionI.cs:619]: (Not from this PR)
Do you know when we can reach this point with _sendStreams.Count == 0?

**Review by bernardnormier — APPROVED** [review]
Looks good.

File: csharp/src/Ice/ConnectionI.cs (hunk 2 above)
Comment (pepone) [reply csharp/src/Ice/ConnectionI.cs:619]: Seems `StartWrite` is some times called with a 0 buffer size just to invoke the completion callback. It is not clear why we do it that way, and cannot call it from the finishAsync.

**Review by externl — APPROVED** [review]

File: csharp/src/Ice/ConnectionI.cs (hunk 1 above)
Comment (bernardnormier) [reply csharp/src/Ice/ConnectionI.cs:616]: Yes, this wait until a follow-up PR.

**Pull request merged**